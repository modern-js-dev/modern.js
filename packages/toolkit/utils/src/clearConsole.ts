export const clearConsole = () => {
  if (process.stdout.isTTY && !process.env.DEBUG) {
    process.stdout.write('\x1B[H\x1B[2J');
  }
};
