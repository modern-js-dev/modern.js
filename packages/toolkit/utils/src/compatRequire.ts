import { findExists } from './findExists';

/**
 * Require function compatible with esm and cjs module.
 * @param filePath - File to required.
 * @returns module export object.
 */
export const compatRequire = (filePath: string) => {
  const mod = require(filePath);

  return mod?.__esModule ? mod.default : mod;
};

export const requireExistModule = (
  filename: string,
  extensions = ['.ts', '.js'],
) => {
  const exist = findExists(extensions.map(ext => `${filename}${ext}`));

  if (!exist) {
    return null;
  }

  return compatRequire(exist);
};

export const cleanRequireCache = (filelist: string[]) => {
  filelist.forEach(filepath => {
    delete require.cache[filepath];
  });
};
