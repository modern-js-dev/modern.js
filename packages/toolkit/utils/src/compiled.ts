import chalk from '../compiled/chalk';
import filesize from '../compiled/filesize';
import stripAnsi from '../compiled/strip-ansi';

export { chalk, filesize, stripAnsi };
