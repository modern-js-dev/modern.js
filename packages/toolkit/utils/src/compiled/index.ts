import chalk from './chalk';
import filesize from './filesize';

export { chalk, filesize };
