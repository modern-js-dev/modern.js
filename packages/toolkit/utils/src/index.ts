export * as fs from 'fs-extra';
export * from './chalk';
export * from './formatWebpackMessages';
export * from './FileSizeReporter';
export * from './printBuildError';
export * from './debug';
export * from './findExists';
export * from './is';
export * from './compatRequire';
export * from './logger';
export * from './constants';
export * from './ensureAbsolutePath';
export * from './getCacheIdentifier';
export * from './clearConsole';
export * from './pkgUp';
export * from './applyOptionsChain';
export * from './getBrowserslist';
export * from './removeSlash';
export * from './getEntryOptions';
export * from './getPort';
export * from './monorepo';
export * from './getPackageManager';
export * from './runtimeExports';
export * from './readTsConfig';
export * from './path';
export * from './generateMetaTags';
export * from './prettyInstructions';
export * from './alias';
export * from './import';
export * from './watch';
export * from './nodeEnv';
