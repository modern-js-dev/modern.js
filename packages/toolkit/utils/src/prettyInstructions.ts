import os from 'os';
import chalk from 'chalk';
import { isDev } from './is';

// TODO: type
interface EntryPoint {
  entryName: string;
}

interface ServerRoute {
  entryName: string;
  isSSR: boolean;
  urlPath: string;
}

// TODO: remove hard code 'main'
export const isSingleEntry = (entrypoints: EntryPoint[]) =>
  entrypoints.length === 1 && entrypoints[0].entryName === 'main';

const normalizeUrl = (url: string) => url.replace(/([^:]\/)\/+/g, '$1');

const getAddressUrls = (protocol = 'http', port: number) => {
  const interfaces = os.networkInterfaces();
  const ipv4Interfaces: os.NetworkInterfaceInfo[] = [];
  Object.keys(interfaces).forEach(key => {
    interfaces[key]!.forEach(detail => {
      if (detail.family === 'IPv4') {
        ipv4Interfaces.push(detail);
      }
    });
  });

  return ipv4Interfaces.reduce(
    (memo: { type: string; url: string }[], detail) => {
      let type = 'Network:  ';
      let url = `${protocol}://${detail.address}:${port}`;
      if (detail.address.includes(`localhost`)) {
        type = 'Local:  ';
        url = `${protocol}://localhost:${port}`;
      }

      memo.push({ type, url });
      return memo;
    },
    [],
  );
};

export const prettyInstructions = (appContext: any, config: any) => {
  const { entrypoints, serverRoutes, port, existSrc, checkedEntries } =
    appContext as {
      entrypoints: EntryPoint[];
      serverRoutes: ServerRoute[];
      port: number;
      existSrc: boolean;
      checkedEntries: string[];
    };

  const urls = getAddressUrls(
    config.dev.https && isDev() ? 'https' : 'http',
    port,
  );

  const routes = existSrc
    ? serverRoutes.filter(route => route.entryName)
    : serverRoutes;

  let message = 'App running at:\n\n';

  if (isSingleEntry(entrypoints) || !existSrc) {
    message += urls
      .map(
        ({ type, url }) =>
          `  ${chalk.bold(`> ${type.padEnd(10)}`)}${chalk.cyanBright(
            normalizeUrl(`${url}/${routes[0].urlPath}`),
          )}\n`,
      )
      .join('');
  } else {
    const maxNameLength = Math.max(...routes.map(r => r.entryName.length));

    urls.forEach(({ type, url }) => {
      message += `  ${chalk.bold(`> ${type}`)}\n`;
      routes.forEach(({ entryName, urlPath, isSSR }) => {
        if (!checkedEntries.includes(entryName)) {
          return;
        }

        message += `    ${chalk.yellowBright(
          isSSR ? 'λ' : '○',
        )}  ${chalk.yellowBright(
          entryName.padEnd(maxNameLength + 8),
        )}${chalk.cyanBright(normalizeUrl(`${url}/${urlPath}`))}\n`;
      });
    });

    message += '\n';
    message += chalk.cyanBright(
      '  λ (Server) server-side renders at runtime\n',
    );
    message += chalk.cyanBright(
      '  ○ (Static) client-side rendered as static HTML\n',
    );
  }

  return message;
};
