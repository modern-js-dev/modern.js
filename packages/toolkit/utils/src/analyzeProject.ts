import * as path from 'path';
import { fs, minimist } from './compiled';

export const isApiOnly = async (appDirectory: string): Promise<boolean> => {
  const srcDir = path.join(appDirectory, 'src');
  const existSrc = await fs.pathExists(srcDir);
  const options = minimist(process.argv.slice(2));
  return !existSrc || Boolean(options['api-only']);
};
