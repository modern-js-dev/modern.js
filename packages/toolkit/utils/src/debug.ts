import { debug } from '../compiled/debug';

/**
 * Create debug function with unified namespace prefix.
 * @param scope - Custom module name of your debug function.
 * @returns Debug function which namespace start with modern-js:.
 */
export const createDebugger = (scope: string) => debug(`modern-js:${scope}`);
