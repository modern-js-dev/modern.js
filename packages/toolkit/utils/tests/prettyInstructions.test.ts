import { prettyInstructions } from '../src/prettyInstructions';

const mockedNetworkInterfaces = {
  lo0: [
    {
      address: '127.0.0.1',
      netmask: '255.0.0.0',
      family: 'IPv4',
      mac: '00:00:00:00:00:00',
      internal: true,
      cidr: '127.0.0.1/8',
    },
  ],
  en5: [
    {
      address: 'fe80::aede:48ff:fe00:1122',
      netmask: 'ffff:ffff:ffff:ffff::',
      family: 'IPv6',
      mac: 'ac:de:48:00:11:22',
      internal: false,
      cidr: 'fe80::aede:48ff:fe00:1122/64',
      scopeid: 4,
    },
  ],
  en0: [
    {
      address: '11.11.111.11',
      netmask: '255.255.252.0',
      family: 'IPv4',
      mac: '90:9c:4a:cf:11:d2',
      internal: false,
      cidr: '10.85.117.60/22',
    },
  ],
  utun2: [
    {
      address: '10.100.100.100',
      netmask: '255.255.224.0',
      family: 'IPv4',
      mac: '00:00:00:00:00:00',
      internal: false,
      cidr: '10.255.182.172/19',
    },
  ],
};

jest.mock('os', () => {
  const originalModule = jest.requireActual('os');
  return {
    __esModule: true,
    ...originalModule,
    default: {
      networkInterfaces() {
        return mockedNetworkInterfaces;
      },
    },
  };
});

jest.mock('chalk', () => ({
  __esModule: true,
  default: {
    bold: jest.fn(str => str),
    green: jest.fn(str => str),
    red: jest.fn(str => str),
    yellow: jest.fn(str => str),
    cyanBright: jest.fn(str => str),
  },
}));

describe('prettyInstructions', () => {
  test('basic usage', () => {
    const mockedAppContext = {
      entrypoints: [
        {
          entryName: 'main',
          entry: '/example/node_modules/.modern-js/main/index.js',
          isAutoMount: true,
          customBootstrap: false,
        },
      ],
      serverRoutes: [
        {
          urlPath: '/',
          entryName: 'main',
          entryPath: 'html/main/index.html',
          isSPA: true,
          isSSR: false,
          enableModernMode: false,
        },
        {
          urlPath: '/api',
          isApi: true,
          entryPath: '',
          isSPA: false,
          isSSR: false,
        },
      ],
      port: 8080,
      existSrc: true,
    };
    const mockedConfig = {
      dev: {
        https: true,
      },
    };

    const message = prettyInstructions(mockedAppContext, mockedConfig);

    expect(message).toMatchSnapshot();
  });

  test('The src directory does not exist', () => {
    const mockedAppContext = {
      entrypoints: [],
      serverRoutes: [
        {
          urlPath: '/api',
          isApi: true,
          entryPath: '',
          isSPA: false,
          isSSR: false,
        },
      ],
      port: 8080,
      existSrc: false,
    };

    const mockedConfig = {
      dev: {
        https: true,
      },
    };

    const message = prettyInstructions(mockedAppContext, mockedConfig);

    expect(message).toMatchSnapshot();
  });
});
