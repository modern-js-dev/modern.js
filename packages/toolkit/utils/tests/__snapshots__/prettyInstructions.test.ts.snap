// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prettyInstructions The src directory does not exist 1`] = `
"App running at:

  > Network:  http://127.0.0.1:8080/api
  > Network:  http://11.11.111.11:8080/api
  > Network:  http://10.100.100.100:8080/api
"
`;

exports[`prettyInstructions basic usage 1`] = `
"App running at:

  > Network:  http://127.0.0.1:8080/
  > Network:  http://11.11.111.11:8080/
  > Network:  http://10.100.100.100:8080/
"
`;
